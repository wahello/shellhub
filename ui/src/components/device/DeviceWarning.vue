<template>
  <fragment>
    <v-dialog
      v-model="show"
      max-width="900"
    >
      <v-card data-test="deviceWarning-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          Update account or select three devices
        </v-card-title>

        <v-card-text class="mt-4 pb-1">
          <p>
            You currently have no subscription to the
            <a :href="url()">
              premium plan
            </a> and the free version is limited to 3 devices. To unlock access to all
            devices, you can subscribe to the
            <a :href="url()">
              premium plan
            </a>. Case, If you want to continue on the free plan, you need to select three devices.
          </p>
        </v-card-text>

        <v-app-bar
          flat
          color="transparent"
          class="mt-0"
        >
          <v-tabs
            centered
          >
            <v-tab
              v-for="item in items"
              :key="item.title"
              :data-test="item.title+'-tab'"
              @click="doAction(item.action)"
            >
              {{ item.title }}
            </v-tab>
          </v-tabs>
        </v-app-bar>

        <v-card-text class="mb-2 pb-0">
          <DeviceListChoice
            :action="action"
            data-test="deviceListChoice-component"
          />
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="close-btn"
            @click="close()"
          >
            Close
          </v-btn>

          <v-tooltip
            :disabled="!showTooltip"
            top
          >
            <template #activator="{ on, attrs }">
              <v-btn
                text
                v-bind="attrs"
                data-test="accept-btn"
                v-on="on"
                @click="accept()"
              >
                Accept
              </v-btn>
            </template>

            <span>
              You need to select 3 devices.
            </span>
          </v-tooltip>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import DeviceListChoice from '@/components/device/DeviceListChoice';

export default {
  name: 'DeviceWarning',

  components: {
    DeviceListChoice,
  },

  data() {
    return {
      hostname: window.location.hostname,
      action: 'suggestedDevices',
      dialog: false,
      items: [
        {
          title: 'Suggested Devices',
          action: 'suggestedDevices',
        },
        {
          title: 'All devices',
          action: 'allDevices',
        },
      ],
    };
  },

  computed: {
    show: {
      get() {
        return this.$store.getters['devices/getDeviceWarning'];
      },

      set(value) {
        this.$store.dispatch('devices/setDeviceWarning', value);
      },
    },

    showTooltip() {
      return this.$store.getters['devices/getDevicesSelected'].length !== 3
        && this.action !== this.items[0].action;
    },

    equalThreeDevices() {
      return this.$store.getters['devices/getDevicesSelected'].length === 3;
    },
  },

  watch: {
    show(status) {
      if (status && this.action === this.items[0].action) {
        this.$store.dispatch('devices/getDevicesMostUsed');
      }
    },
  },

  methods: {
    async doAction(action) {
      this.action = action;

      if (action === this.items[0].action) {
        this.$store.dispatch('devices/getDevicesMostUsed');
      } else if (action === this.items[1].action) {
        const data = {
          perPage: 10,
          page: 1,
          filter: this.$store.getters['devices/getFilter'],
          status: 'accepted',
          sortStatusField: null,
          sortStatusString: 'asc',
        };

        try {
          await this.$store.dispatch('devices/fetch', data);
        } catch (error) {
          if (error.response.status === 403) {
            this.$store.dispatch('snackbar/showSnackbarErrorAssociation');
          } else {
            this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceList);
          }
        }
      } else {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.deviceList);
      }
    },

    accept() {
      if (this.action === this.items[0].action) {
        this.sendDevicesChoice(this.$store.getters['devices/list']);
      } else if (this.$store.getters['devices/getDevicesSelected'].length === 3) {
        this.sendDevicesChoice(this.$store.getters['devices/getDevicesSelected']);
      } else {
        this.$store.dispatch('snackbar/showSnackbarDeviceChoice');
      }
    },

    async sendDevicesChoice(devices) {
      const choices = [];
      devices.forEach((device) => {
        choices.push(device.uid);
      });

      try {
        await this.$store.dispatch('devices/postDevicesChoice', { choices });
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.deviceChoice);

        this.close();
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.deviceChoice);
      }

      this.$store.dispatch('stats/get');
    },

    url() {
      return `${window.location.protocol}//${this.hostname}/settings/billing`;
    },

    close() {
      this.$store.dispatch('devices/setDeviceWarning', false);
    },
  },
};

</script>
