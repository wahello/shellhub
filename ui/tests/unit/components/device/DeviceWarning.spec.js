import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import DeviceWarning from '@/components/device/DeviceWarning';

describe('DeviceWarning', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  let wrapper;

  const dialog = false;
  const deviceWarning = false;
  const devicesSelected = [];

  const hostname = 'localhost';
  const url = `http://${hostname}/settings/billing`;

  const filter = [];

  const items = [
    {
      title: 'Suggested Devices',
      action: 'suggestedDevices',
    },
    {
      title: 'All devices',
      action: 'allDevices',
    },
  ];

  const devices = [
    {
      uid: 'a582b47a42d',
      name: '39-5e-2a',
      identity: {
        mac: '00:00:00:00:00:00',
      },
      info: {
        id: 'linuxmint',
        pretty_name: 'Linux Mint 19.3',
        version: '',
      },
      public_key: '----- PUBLIC KEY -----',
      tenant_id: '00000000',
      last_seen: '2020-05-20T18:58:53.276Z',
      online: false,
      namespace: 'user',
      status: 'accepted',
    },
    {
      uid: 'a582b47a42e',
      name: '39-5e-2b',
      identity: {
        mac: '00:00:00:00:00:00',
      },
      info: {
        id: 'linuxmint',
        pretty_name: 'Linux Mint 19.3',
        version: '',
      },
      public_key: '----- PUBLIC KEY -----',
      tenant_id: '00000001',
      last_seen: '2020-05-20T19:58:53.276Z',
      online: true,
      namespace: 'user',
      status: 'accepted',
    },
  ];

  const store = new Vuex.Store({
    namespaced: true,
    state: {
      deviceWarning,
      devicesSelected,
      filter,
      devices,
    },
    getters: {
      'devices/getDeviceWarning': (state) => state.deviceWarning,
      'devices/getDevicesSelected': (state) => state.devicesSelected,
      'devices/getFilter': (state) => state.filter,
      'devices/list': (state) => state.devices,
    },
    actions: {
      'devices/getDevicesMostUsed': () => {},
      'devices/fetch': () => {},
      'snackbar/showSnackbarDeviceChoice': () => {},
      'devices/postDevicesChoice': () => {},
      'stats/get': () => {},
      'devices/setDeviceWarning': () => {},
      'snackbar/showSnackbarErrorAssociation': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
    },
  });

  describe('Dialog is closes', () => {
    beforeEach(() => {
      wrapper = mount(DeviceWarning, {
        store,
        localVue,
        stubs: ['fragment'],
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      document.body.setAttribute('data-app', true);
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.hostname).toEqual(hostname);
      expect(wrapper.vm.action).toEqual(items[0].action);
      expect(wrapper.vm.items).toEqual(items);
      expect(wrapper.vm.dialog).toEqual(dialog);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.show).toEqual(false);
      expect(wrapper.vm.showTooltip).toEqual(false);
      expect(wrapper.vm.equalThreeDevices).toEqual(false);
    });
    it('Process data in methods', () => {
      expect(wrapper.vm.url()).toEqual(url);

      wrapper.vm.close();
      expect(wrapper.vm.show).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', async () => {
      expect(wrapper.find('[data-test="deviceListChoice-component"]').exists()).toEqual(false);
    });
    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="deviceWarning-dialog"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="accept-btn"]').exists()).toEqual(false);
    });
  });
});
