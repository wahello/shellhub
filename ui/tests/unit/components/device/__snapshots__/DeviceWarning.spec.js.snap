// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceWarning Dialog is closes Renders the component 1`] = `
<fragment-stub>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;
